server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/cache/CachingRepository.java

```diff
@@ -272,8 +323,64 @@ public RevisionRange normalizeNow(Revision from, Revision to) {
         return repo.commit(baseRevision, commitTimeMillis, author, summary, detail, markup, changes);
     }
 
-    private <T> CompletableFuture<T> normalizeAndCompose(
-            Revision revision, Function<Revision, CompletableFuture<T>> function) {
-        return normalize(revision).thenCompose(function);
+    @Override
+    public <T> CompletableFuture<MergedEntry<T>> mergeFiles(Revision revision, MergeQuery<T> query) {
+        requireNonNull(revision, "revision");
+        requireNonNull(query, "query");
+
+        final Revision normalizedRevision;
+        try {
+            normalizedRevision = normalizeNow(revision);
+        } catch (Exception e) {
+            return CompletableFutures.exceptionallyCompletedFuture(e);
+        }
+
+        final CacheableMergeQueryCall key = new CacheableMergeQueryCall(repo, normalizedRevision, query);
+        final CompletableFuture<Object> value = cache.getIfPresent(key);
+        if (value != null) {
+            return unsafeCast(value);
+        }
+
+        final List<MergeSource> mergeSources = query.mergeSources();
```

Comment:
I'm curious what if we just call `Repository.mergeFiles()` here, instead of L344-375, like as follows:
```
return Repository.super.mergeFiles(revision, query).thenApply(mergedEntry -> {
    cache.get(key, unused -> mergedEntry);
    return mergedEntry;
});
```

Reply:
I realized that this may be related to @trustin 's comment: https://github.com/line/centraldogma/pull/293#discussion_r236561229.
Forget this comment if you can resolve his comment. :-)